src/xarray_regrid/methods/conservative.py

```diff
@@ -102,132 +114,88 @@ def conservative_regrid_dataset(
-            target_coords = coords[coord].to_numpy()
-            source_coords = data[coord].to_numpy()
+    for array, attrs in data_attrs.items():
+        data_vars[array].attrs = attrs
 
-            weights = get_weights(source_coords, target_coords)
+    ds_regridded = xr.Dataset(data_vars=data_vars, attrs=ds_attrs)
 
-            # Modify weights to correct for latitude distortion
-            if str(coord) == latitude_coord:
-                dot_array = utils.create_dot_dataarray(
-                    weights, str(coord), target_coords, source_coords
-                )
-                dot_array = apply_spherical_correction(dot_array, latitude_coord)
-                weights = dot_array.to_numpy()
+    for coord, attrs in coord_attrs.items():
+        if coord not in ds_regridded.coords:
+            # Add back any additional coordinates from the original dataset
+            ds_regridded[coord] = data_coords[coord]
+        ds_regridded[coord].attrs = attrs
 
-            data = data.transpose(coord, ...)
-            data = apply_weights(data, weights, coord, target_coords)
+    return ds_regridded
 
-            # Replace zeros outside of original data grid with NaNs
-            data = data.where(uncovered_target_grid)
 
-    new_coords = [data[coord] for coord in data_coords]
-    for coord, attr in zip(new_coords, coord_attrs, strict=True):
-        coord.attrs = attr
-    data.attrs = attrs
+def apply_weights(
+    da: xr.DataArray,
+    weights: np.ndarray,
+    coord_name: Hashable,
+    valid_frac: xr.DataArray | None,
+    skipna: bool,
+) -> tuple[xr.DataArray, xr.DataArray]:
+    """Apply the weights to convert data to the new coordinates."""
+    coord_map = {f"target_{coord_name}": coord_name}
+    weights_norm = weights.copy()
 
-    return data
+    if skipna:
+        notnull = da.notnull()
+        da = da.fillna(0)
+        # Renormalize the weights along this dim by the accumulated valid_frac
+        # along previous dimensions
+        if valid_frac is not None:
+            weights_norm = weights * valid_frac / valid_frac.mean(coord_name)
 
+    da_reduced = xr.dot(da, weights_norm, dim=coord_name, optimize=True)
+    da_reduced = da_reduced.rename(coord_map).transpose(*da.dims)
 
-def apply_weights(
-    da: xr.DataArray, weights: np.ndarray, coord_name: Hashable, new_coords: np.ndarray
-) -> xr.DataArray:
-    """Apply the weights to convert data to the new coordinates."""
-    new_data: np.ndarray | dask.array.Array
-    if da.chunks is not None:
-        # Dask routine
-        new_data = compute_einsum_dask(da, weights)
-    else:
-        # numpy routine
-        new_data = compute_einsum_numpy(da, weights)
-
-    coord_mapping = {coord_name: new_coords}
-    coords = list(da.dims)
-    coords.remove(coord_name)
-    for coord in coords:
-        coord_mapping[coord] = da[coord].to_numpy()
+    if skipna:
+        weights_valid_sum = xr.dot(
+            weights_norm, notnull, dim=coord_name, optimize=True
```

Comment:
We can lessen the performance penalty significantly by doing something like `notnull.any([non_grid_dims])` here, in which case we track the nan fraction as any batch slices that have valid data. Could be a reasonable tradeoff, or a configurable argument.

Reply:
Could you open an issue on this? It's probably nicer to tackle that separately and round off this PR first